File: java/src/Ice/src/main/java/com/zeroc/IceInternal/RequestHandlerFactory.java
// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package com.zeroc.IceInternal;

import java.util.Map;
import java.util.HashMap;
import java.util.concurrent.Callable;

public final class RequestHandlerFactory
{
    RequestHandlerFactory(Instance instance)
    {
        _instance = instance;
    }

    public RequestHandler
    getRequestHandler(final RoutableReference ref, com.zeroc.Ice._ObjectPrxI proxy)
    {
        if(ref.getCollocationOptimized())
        {
            com.zeroc.Ice.ObjectAdapter adapter = _instance.objectAdapterFactory().findObjectAdapter(proxy);
            if(adapter != null)
            {
                return proxy.__setRequestHandler(new CollocatedRequestHandler(ref, adapter));
            }
        }

        ConnectRequestHandler handler = null;
        boolean connect = false;
        if(ref.getCacheConnection())
        {
            synchronized(this)
            {
                handler = _handlers.get(ref);
                if(handler == null)
                {
                    handler = new ConnectRequestHandler(ref, proxy);
                    _handlers.put(ref, handler);
                    connect = true;
                }
            }
        }
        else
        {
            handler = new ConnectRequestHandler(ref, proxy);
            connect = true;
        }

        if(connect)
        {
            if(_instance.queueRequests())
            {
                final ConnectRequestHandler h = handler;
                _instance.getQueueExecutor().executeNoThrow(new Callable<Void>()
                                                            {
                                                                @Override
                                                                public Void call()
                                                                {
                                                                    ref.getConnection(h);
                                                                    return null;
                                                                }
                                                            });
            }
            else
            {
                ref.getConnection(handler);
            }
        }
        return proxy.__setRequestHandler(handler.connect(proxy));
    }

    void
    removeRequestHandler(Reference ref, RequestHandler handler)
    {
        if(ref.getCacheConnection())
        {
            synchronized(this)
            {
                if(_handlers.get(ref) == handler)
                {
                    _handlers.remove(ref);
                }
            }
        }
    }

    private final Instance _instance;
    private final Map<Reference, ConnectRequestHandler> _handlers = new HashMap<>();
}

File: csharp/src/Ice/Internal/RequestHandlerFactory.cs
// Copyright (c) ZeroC, Inc.

namespace Ice.Internal;

public class RequestHandlerFactory
{
    internal RequestHandlerFactory(Instance instance)
    {
        _instance = instance;
    }

    public RequestHandler
    getRequestHandler(RoutableReference rf, Ice.ObjectPrxHelperBase proxy)
    {
        if (rf.getCollocationOptimized())
        {
            Ice.ObjectAdapter adapter = _instance.objectAdapterFactory().findObjectAdapter(rf);
            if (adapter != null)
            {
                return proxy.iceSetRequestHandler(new CollocatedRequestHandler(rf, adapter));
            }
        }

        bool connect = false;
        ConnectRequestHandler handler;
        if (rf.getCacheConnection())
        {
            lock (this)
            {
                if (!_handlers.TryGetValue(rf, out handler))
                {
                    handler = new ConnectRequestHandler(rf);
                    _handlers.Add(rf, handler);
                    connect = true;
                }
            }
        }
        else
        {
            handler = new ConnectRequestHandler(rf);
            connect = true;
        }

        if (connect)
        {
            rf.getConnection(handler);
        }
        return proxy.iceSetRequestHandler(handler.connect(proxy));
    }

    internal void
    removeRequestHandler(Reference rf, RequestHandler handler)
    {
        if (rf.getCacheConnection())
        {
            lock (this)
            {
                ConnectRequestHandler h;
                if (_handlers.TryGetValue(rf, out h) && h == handler)
                {
                    _handlers.Remove(rf);
                }
            }
        }
    }

    private readonly Instance _instance;
    private readonly Dictionary<Reference, ConnectRequestHandler> _handlers = new();
}
